### PR TITLE
Merge branch 'users/sschen/wip' into upstream

### PR DESCRIPTION
**Pull request by HSAR**
This is a squashed commit. Adds PMD support to the Maven build task.

### CODE REVIEW

**Comment by HSAR** [comment]
There are no code changes from the previous commit(s).

File: Tasks/Maven/mavencodeanalysis.ts [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        // Put the artifacts in subdirectories corresponding to their module
+        containerfolder: moduleAnalysis.moduleName,
+        // Prefix the build number onto the upload for convenience
+        // NB: Artifact names need to be unique on an upload-by-upload basis
+        artifactname: buildNumber + '_' + moduleAnalysis.moduleName + '_' + toolName
+    }, localStagingDir);
+}
+
+// Extract data from code analysis output files and upload results to build server
+export function uploadCodeAnalysisResults():void {
```
Comment (bgavrilMS) [inline Tasks/Maven/mavencodeanalysis.ts]: I would have the "public" functions before the "private" functions to make the file easier to read, but I don't feel strongly about it.

**Comment by bgavrilMS** [comment]
LGTM

File: Tasks/Maven/mavencodeanalysis.ts (hunk 1 above)
Comment (HSAR) [inline Tasks/Maven/mavencodeanalysis.ts]: Accepted, done.

The file isn't as messy as I thought, and so I just moved those two entry points to the top.

Comment (HSAR) [inline Tasks/Maven/mavencodeanalysis.ts]: Resolved.

**Pull request merged**